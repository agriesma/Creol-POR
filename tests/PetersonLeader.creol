/*
 * PetersonLeader.creol -- LeaderElection inspired by Peterson.
 *
 * This file is part of creoltools
 *
 * Written and Copyright (c) 2007 by Marcel Kyas
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License as
 * published by the Free Software Foundation; either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

interface Process
begin
with Process
  op send(in u: Int)
end

class Process(i: Int, neighbour: Process) contracts Process
begin

  var mode: Bool := true // true is active, false is relay
  var status: Int := 0   // 0 is unknown, 1 is chosen, 2 is reported
  var uid1: Int := i
  var uid2: Int
  var uid3: Int

  op init == !neighbour.send(i)

  with Process op send(in u: Int) ==
    if mode then
      if uid2 = null then
        uid2 := u;
        !neighbour.send(uid2);
        if uid2 = uid1 then status := 1 end
      else
        assert uid3 = null;
        uid3 := u;
        if (uid2 > uid1 && uid2 > uid3) then
          // Advance phase
          uid1, uid2, uid3 := uid2, null, null;
          !neighbour.send(uid1)
        else
          // Become relay
          mode := false
        end
      end
    else
      // Relay
      !neighbour.send(u)
    end

end
