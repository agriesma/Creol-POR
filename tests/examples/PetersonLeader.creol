/*
 * PetersonLeader.creol -- LeaderElection inspired by Peterson.
 *
 * This file is part of creoltools
 *
 * Written and Copyright (c) 2007 by Marcel Kyas
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License as
 * published by the Free Software Foundation; either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

interface Main
begin
with Process
  op getNeighbour(out n: Process)
end

interface Process
begin
with Process
  op send(in u: Int)
end

class Process(i: Int, main: Main) contracts Process
begin
  var neighbour: Process
  var mode: Bool := true // true is active, false is relay
  var status: Int := 0   // 0 is unknown, 1 is chosen, 2 is reported
  var uid1: Int := i
  var uid2: Int

  op init == main.getNeighbour(; neighbour); !neighbour.send(i)

  with Process op send(in u: Int) ==
    if mode then
      if uid2 = null then
        if u = uid1 then status := 1 else uid2 := u; !neighbour.send(u) end
      else
        if (uid2 > uid1 && uid2 > u) then
          // Advance phase
          uid1, uid2 := uid2, null; !neighbour.send(uid1)
        else
          // Become relay
          mode := false
        end
      end
    else
      // Relay
      !neighbour.send(u)
    end

end

class Main contracts Main
begin
  var p1: Process
  var p2: Process
  var p3: Process
  var p4: Process
  var p5: Process

  op init == p1 := new Process(1, this); p2 := new Process(3, this);
    p3 := new Process(5, this); p4 := new Process(4, this);
    p5 := new Process(2, this)

  with Process op getNeighbour(out n: Process) ==
    if caller = p1 then n := p2 else
    if caller = p2 then n := p3 else
    if caller = p3 then n := p4 else
    if caller = p4 then n := p5 else
    if caller = p5 then n := p1 end end end end end
end
