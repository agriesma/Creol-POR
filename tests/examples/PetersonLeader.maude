load creol-interpreter
mod PROGRAM is
protecting CREOL-SIMULATOR .
op init : -> Configuration [ctor] .
eq init =
< "Process" : Cl | Inh: noInh, Par: "i" , "main", Att: "i" |-> null , "main" |-> null , "neighbour" |-> null , "mode" |-> null , "status" |-> null , "uid1" |-> null , "uid2" |-> null, Mtds: 
  < "run" : Mtdname | Param: noVid, Latt: noSubst, Code: skip ; return ( emp ) > *
  < "init" : Mtdname | Param: noVid, Latt: "label:1" |-> null , "label:0" |-> null, Code:  ( ( "mode" @@ "Process" ), ( "status" @@ "Process" ), ( "uid1" @@ "Process" ) )  ::= bool(true) # int(0) # "i" ; "label:0" ! "main" . "getNeighbour" ( emp ) ; ( "label:0" ? ( "neighbour" ) )  ; "label:1" ! "neighbour" . "send" ( "i" ) ; free( "label:1" ) ; return ( emp ) > *
  < "send" : Mtdname | Param: "u", Latt: "u" |-> null , "label:2" |-> null , "label:1" |-> null , "label:0" |-> null, Code: if "mode" th if "=" ( "uid2" # null ) th if "=" ( "u" # "uid1" ) th "status" ::= int(1) el "uid2" ::= "u" ; "label:0" ! "neighbour" . "send" ( "u" ) ; free( "label:0" ) fi el if "&&" ( ">" ( "uid2" # "uid1" ) # ">" ( "uid2" # "u" ) ) th  ( "uid1", "uid2" )  ::= "uid2" # null ; "label:1" ! "neighbour" . "send" ( "uid1" ) ; free( "label:1" ) el "mode" ::= bool(false) fi fi el "label:2" ! "neighbour" . "send" ( "u" ) ; free( "label:2" ) fi ; return ( emp ) >, Ocnt: 0 >

< "Main" : Cl | Inh: noInh, Par: noVid, Att: "p1" |-> null , "p2" |-> null , "p3" |-> null , "p4" |-> null , "p5" |-> null, Mtds: 
  < "run" : Mtdname | Param: noVid, Latt: noSubst, Code: skip ; return ( emp ) > *
  < "init" : Mtdname | Param: noVid, Latt: noSubst, Code: "p1" ::= new "Process" ( int(1) # "this" ) ; "p2" ::= new "Process" ( int(3) # "this" ) ; "p3" ::= new "Process" ( int(5) # "this" ) ; "p4" ::= new "Process" ( int(4) # "this" ) ; "p5" ::= new "Process" ( int(2) # "this" ) ; return ( emp ) > *
  < "getNeighbour" : Mtdname | Param: noVid, Latt: "n" |-> null, Code: if "=" ( "caller" # "p1" ) th "n" ::= "p2" el if "=" ( "caller" # "p2" ) th "n" ::= "p3" el if "=" ( "caller" # "p3" ) th "n" ::= "p4" el if "=" ( "caller" # "p4" ) th "n" ::= "p5" el if "=" ( "caller" # "p5" ) th "n" ::= "p1" el skip fi fi fi fi fi ; return ( "n" ) >, Ocnt: 0 >

.
endm
