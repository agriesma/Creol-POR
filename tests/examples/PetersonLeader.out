==========================================
reduce in PROGRAM : init .
result Configuration: 
< "Main" : Cl | Inh: noInh, Par: noVid, Att: "p1" |-> null, "p2" |-> null, "p3"
    |-> null, "p4" |-> null, "p5" |-> null, Mtds:
  < "getNeighbour" : Mtdname | Param: noVid, Latt: "n" |-> null, Code: if "="(
    "caller" # "p1") th "n" ::= "p2" el if "="("caller" # "p2") th "n" ::= "p3"
    el if "="("caller" # "p3") th "n" ::= "p4" el if "="("caller" # "p4") th
    "n" ::= "p5" el if "="("caller" # "p5") th "n" ::= "p1" el skip fi fi fi fi
    fi ; return("n") > *
  < "init" : Mtdname | Param: noVid, Latt: noSubst, Code: "p1" ::= new
    "Process"(int(1) # "this") ; "p2" ::= new "Process"(int(3) # "this") ; "p3"
    ::= new "Process"(int(5) # "this") ; "p4" ::= new "Process"(int(4) #
    "this") ; "p5" ::= new "Process"(int(2) # "this") ; return(emp) > *
  < "run" : Mtdname | Param: noVid, Latt: noSubst, Code: skip ; return(emp) >,
  Ocnt: 0 >

< "Process" : Cl | Inh: noInh, Par: "i", "main", Att: "i" |-> null, "main" |->
    null, "mode" |-> null, "neighbour" |-> null, "status" |-> null, "uid1" |->
    null, "uid2" |-> null, Mtds:
  < "init" : Mtdname | Param: noVid, Latt: "label:0" |-> null, "label:1" |->
    null, Code: (("mode" @@ "Process"), ("status" @@ "Process"), ("uid1" @@
    "Process")) ::= bool(true) # int(0) # "i" ; "label:0" ! "main" .
    "getNeighbour"(emp) ; "label:0" ?("neighbour") ; "label:1" ! "neighbour" .
    "send"("i") ; free("label:1") ; return(emp) > *
  < "run" : Mtdname | Param: noVid, Latt: noSubst, Code: skip ; return(emp) > *
  < "send" : Mtdname | Param: "u", Latt: "label:0" |-> null, "label:1" |->
    null, "label:2" |-> null, "u" |-> null, Code: if "mode" th if "="("uid2" #
    null) th if "="("u" # "uid1") th "status" ::= int(1) el "uid2" ::= "u" ;
    "label:0" ! "neighbour" . "send"("u") ; free("label:0") fi el if "&&"(">"(
    "uid2" # "uid1") # ">"("uid2" # "u")) th ("uid1", "uid2") ::= "uid2" # null
    ; "label:1" ! "neighbour" . "send"("uid1") ; free("label:1") el "mode" ::=
    bool(false) fi fi el "label:2" ! "neighbour" . "send"("u") ; free(
    "label:2") fi ; return(emp) >,
  Ocnt: 0 >
==========================================
rewrite in PROGRAM : init
main("Main", emp) .
result Configuration: 
< "Main" : Cl | Inh: noInh, Par: noVid, Att: "p1" |-> null, "p2" |-> null, "p3"
    |-> null, "p4" |-> null, "p5" |-> null, Mtds:
  < "getNeighbour" : Mtdname | Param: noVid, Latt: "n" |-> null, Code: if "="(
    "caller" # "p1") th "n" ::= "p2" el if "="("caller" # "p2") th "n" ::= "p3"
    el if "="("caller" # "p3") th "n" ::= "p4" el if "="("caller" # "p4") th
    "n" ::= "p5" el if "="("caller" # "p5") th "n" ::= "p1" el skip fi fi fi fi
    fi ; return("n") > *
  < "init" : Mtdname | Param: noVid, Latt: noSubst, Code: "p1" ::= new
    "Process"(int(1) # "this") ; "p2" ::= new "Process"(int(3) # "this") ; "p3"
    ::= new "Process"(int(5) # "this") ; "p4" ::= new "Process"(int(4) #
    "this") ; "p5" ::= new "Process"(int(2) # "this") ; return(emp) > *
  < "run" : Mtdname | Param: noVid, Latt: noSubst, Code: skip ; return(emp) >,
  Ocnt: 1 >

< "Process" : Cl | Inh: noInh, Par: "i", "main", Att: "i" |-> null, "main" |->
    null, "mode" |-> null, "neighbour" |-> null, "status" |-> null, "uid1" |->
    null, "uid2" |-> null, Mtds:
  < "init" : Mtdname | Param: noVid, Latt: "label:0" |-> null, "label:1" |->
    null, Code: (("mode" @@ "Process"), ("status" @@ "Process"), ("uid1" @@
    "Process")) ::= bool(true) # int(0) # "i" ; "label:0" ! "main" .
    "getNeighbour"(emp) ; "label:0" ?("neighbour") ; "label:1" ! "neighbour" .
    "send"("i") ; free("label:1") ; return(emp) > *
  < "run" : Mtdname | Param: noVid, Latt: noSubst, Code: skip ; return(emp) > *
  < "send" : Mtdname | Param: "u", Latt: "label:0" |-> null, "label:1" |->
    null, "label:2" |-> null, "u" |-> null, Code: if "mode" th if "="("uid2" #
    null) th if "="("u" # "uid1") th "status" ::= int(1) el "uid2" ::= "u" ;
    "label:0" ! "neighbour" . "send"("u") ; free("label:0") fi el if "&&"(">"(
    "uid2" # "uid1") # ">"("uid2" # "u")) th ("uid1", "uid2") ::= "uid2" # null
    ; "label:1" ! "neighbour" . "send"("uid1") ; free("label:1") el "mode" ::=
    bool(false) fi fi el "label:2" ! "neighbour" . "send"("u") ; free(
    "label:2") fi ; return(emp) >,
  Ocnt: 5 >

< ob("Main0") : "Main" |
  Att: "p1" |-> ob("Process0"), "p2" |-> ob("Process1"), "p3" |-> ob(
    "Process2"), "p4" |-> ob("Process3"), "p5" |-> ob("Process4"), "this" |->
    ob("Main0"),
  Pr: idle,
  PrQ: noProc,
  Dealloc: noDealloc,
  Ev: noMsg,
  Lcnt: 2 >

< ob("Process0") : "Process" |
  Att: "i" |-> int(1), "main" |-> ob("Main0"), "mode" |-> bool(false),
    "neighbour" |-> ob("Process1"), "status" |-> int(0), "this" |-> ob(
    "Process0"), "uid1" |-> int(1), "uid2" |-> int(2),
  Pr: idle,
  PrQ: noProc,
  Dealloc: noDealloc,
  Ev: noMsg,
  Lcnt: 6 >

< ob("Process1") : "Process" |
  Att: "i" |-> int(3), "main" |-> ob("Main0"), "mode" |-> bool(false),
    "neighbour" |-> ob("Process2"), "status" |-> int(0), "this" |-> ob(
    "Process1"), "uid1" |-> int(3), "uid2" |-> int(1),
  Pr: idle,
  PrQ: noProc,
  Dealloc: noDealloc,
  Ev: noMsg,
  Lcnt: 6 >

< ob("Process2") : "Process" |
  Att: "i" |-> int(5), "main" |-> ob("Main0"), "mode" |-> bool(false),
    "neighbour" |-> ob("Process3"), "status" |-> int(0), "this" |-> ob(
    "Process2"), "uid1" |-> int(5), "uid2" |-> int(3),
  Pr: idle,
  PrQ: noProc,
  Dealloc: noDealloc,
  Ev: noMsg,
  Lcnt: 6 >

< ob("Process3") : "Process" |
  Att: "i" |-> int(4), "main" |-> ob("Main0"), "mode" |-> bool(true),
    "neighbour" |-> ob("Process4"), "status" |-> int(1), "this" |-> ob(
    "Process3"), "uid1" |-> int(5), "uid2" |-> null,
  Pr: idle,
  PrQ: noProc,
  Dealloc: noDealloc,
  Ev: noMsg,
  Lcnt: 6 >

< ob("Process4") : "Process" |
  Att: "i" |-> int(2), "main" |-> ob("Main0"), "mode" |-> bool(false),
    "neighbour" |-> ob("Process0"), "status" |-> int(0), "this" |-> ob(
    "Process4"), "uid1" |-> int(2), "uid2" |-> int(4),
  Pr: idle,
  PrQ: noProc,
  Dealloc: noDealloc,
  Ev: noMsg,
  Lcnt: 6 >

< ob("main") : "" |
  Att: noSubst,
  Pr: idle,
  PrQ: noProc,
  Dealloc: noDealloc,
  Ev: noMsg,
  Lcnt: 0 >
Bye.
